### DISCUSSION TITLE
客户端唯一性判断

### DISCUSSION THREAD

**Discussion by zhaohfup** [apolloconfig/apollo#4154] [Q&A]
一个tomcat下部署多个war包，每个war包都有一个app.id，但实际只有第一个启动的war包链接上了apollo，其他war包使用的配置全部来源于第一个启动的war服务。问题是这种机制原理是什么？

**Comment by nobodyiam** [comment]
You may take a look at [DefaultApplicationProvider](https://github.com/apolloconfig/apollo/blob/master/apollo-core/src/main/java/com/ctrip/framework/foundation/internals/provider/DefaultApplicationProvider.java).

**Reply by zhaohfup** [reply]
像我们这种情况多个app.id同时生效怎么配置？

**Reply by Anilople** [reply]
see https://www.apolloconfig.com/#/zh/usage/java-sdk-user-guide

source code 

https://github.com/apolloconfig/apollo/blob/master/apollo-core/src/main/java/com/ctrip/framework/foundation/internals/provider/DefaultApplicationProvider.java#L119-L148

Maybe you can write app.id to your own `META-INF/app.properties`.

**Reply by iambiglee** [reply]
![image](https://user-images.githubusercontent.com/40035800/146470938-00ea41f5-c584-4580-8f67-cb80c288502e.png)
你用的应该是第三种方法，第三种不支持war包的形式，可以采用第四种方式

**Reply by iambiglee** [reply]
多个war包只有一个appid生效的原理如下：
[ApolloApplicationContextInitializer](https://github.com/apolloconfig/apollo/blob/aa9b7ebd01f860c20df9c14129dbbc35cb9d989c/apollo-client/src/main/java/com/ctrip/framework/apollo/spring/boot/ApolloApplicationContextInitializer.java)，实现了EnvironmentPostProcessor接口，该接口中有个初始化方法。通过改初始化方法，最终调用方法[fillSystemPropertyFromEnvironment](https://github.com/apolloconfig/apollo/blob/aa9b7ebd01f860c20df9c14129dbbc35cb9d989c/apollo-client/src/main/java/com/ctrip/framework/apollo/spring/boot/ApolloApplicationContextInitializer.java#L166)，会将配置文件中的["app.id", "apollo.cluster", "apollo.cacheDir", "apollo.meta"]放到System中去，一个JVM线程只有一个System.这就导致了一个tomcat启动后，app.id 会被一直覆盖，所以只有一个app id 生效。

**Comment by iambiglee** [comment]
By the way a question.
Why use `System `to store `app.id` and staff ?
In the case of multiple war packages in a tomcat, This will cause app.id to overwrite each other.
Would it be better if use custom local `ConCurrentHashMap` to store it?
Or there are some specific situation that I didn’t think about.
Tks.

**Reply by nobodyiam** [reply]
Apollo uses System Property to store `app.id` to ease the configuration, e.g. user could pass the `app.id` as `-Dapp.id=xxx`.
The use of `ApolloApplicationContextInitializer` is to ease the configuration for spring boot, e.g. user could set the `app.id` in the `application.properties`.
However, if users set the `app.id` in `META-INF/app.properties`, it won't be stored to System Property, which I think would work in your scenario.

**Reply by iambiglee** [reply]
Got it, Tks
